File: vademecum/code-src/podstawy/Lambda.cpp

/*
C++ pozwala także na definiowanie i używanie lambd. Definicja taka składa
się z listy przechwytywanych zmiennych, listy argumentów i ciała funkcji.

Lista przechwytywania może określać przechwytywanie przez wartość lub przez referencję.

W pierwszym przypadku wartość zmiennej z miejsca utworzenia funkcji zostanie w niej "zamrożona",
czyli jej dalsze zmiany nie będą widoczne w wywołaniach lambdy.

W drugim przypadku lambda będzie widzieć zawsze aktualną wartość,
a zmiany tej zmiennej wewnątrz lambdy będą widoczne także na zewnątrz.

Lista argumentów i ciało funkcji działa jak w zwykłych funkcjach.
Lambda może zwracać lub może nie zwracać wartość z użyciem return.
*/

#include <iostream>
int main() {
	int x = 10, z = 3;
	
	// lambda bez argumentów i listy przechwytywania
	auto lambda_function1 = []() {
		std::cout << "Hello\n";
	};
	lambda_function1();
	
	// lambda z argumentami
	auto lambda_function2 = [](int y) {
		std::cout << "y = " << y << "\n";
	};
	lambda_function2(11);
	
	// lambda z listą przechwytywania (przez wartość)
	// x jest zamrożone jako 10
	auto lambda_function3 = [x](int y) { return x + y; };
	std::cout << lambda_function3(3) << "\n";
	x = 0;
	std::cout << lambda_function3(3) << "\n";
	
	// lambda z listą przechwytywania (przez referencje)
	// odwołujemy się do zewnętrznego x
	auto lambda_function4 = [&x](int y) { return x + y; };
	std::cout << lambda_function4(3) << "\n";
	x = 10;
	std::cout << lambda_function4(3) << "\n";
	
	// w listach przechwytywania można podać wiele zmiennych
	// można mieszać przechwytywanie przez wartość
	// i referencję w ramach jednej lambdy
	
	auto lambda_function5 = [&x, z](int y) { return z * x + y; };
	
	// w C++ lambda może korzystać tylko ze zmiennych określonych
	// w jej listach przechwytywania i argumentów, czyli:
	//   auto lambda_function6 = [](int y) { return x + y; };
	// nie będzie poprawne
}

File: vademecum/code-src/podstawy/Lambda.py

# Python pozwala na definiowanie funkcji wewnątrz funkcji oraz łatwe przechowywanie
# funkcji w zmiennych, także kolekcjach takich jak listy i słowniki.
# Warto zwrócić uwagę na sposób obsługi zmiennych globalnych,
# czy też zewnętrznych w takich przypadkach.

# W poniższym przykładzie widzimy, że funkcje tak zdefiniowane nie korzystają
# z wartości zmiennej x z miejsca definicji, tylko z miejsca wywołania.
# Jeżeli potrzebujemy aby funkcja używała wartości z chwili definicji to możemy
# użyć x jako wartości domyślnej któregoś z argumentów tej funkcji.

tablicaA = [None] * 4
tablicaB = [None] * 4

for x in [0, 1, 2, 3]:
	def tmpA(a):
		return x+a
	def tmpB(a, x=x): # x=x nakazuje przechwycić obecną wartość x
		return x+a
	tablicaA[x]=tmpA
	tablicaB[x]=tmpB

print (x, " + 3 =", tablicaA[1](3), "    1 + 3 = ", tablicaB[1](3))
x = 0
print (x, " + 3 =", tablicaA[1](3), "    1 + 3 = ",  tablicaB[1](3))


# Lambda jest innym sposobem lokalnego definiowania funkcji (często dostępnym także
# w językach nie pozwalających na tak swobodne definiowanie zwykłych funkcji)
# W tym przypadku definicja jest jeszcze bardziej lokalna,
# funkcja nawet nie posiada swojej nazwy.
# Lambda może być przypisana do jakiejś zmiennej lub przekazana w argumencie.
# Sposób traktowania zmiennych zewnętrznych jest analogiczny.

tablicaA = [None] * 4
tablicaB = [None] * 4

for x in [0, 1, 2, 3]:
	tablicaA[x] = lambda a: x+a
	tablicaB[x] = lambda a,x=x: x+a # x=x nakazuje przechwycić obecną wartość x

print (x, " + 3 =", tablicaA[1](3), "    1 + 3 = ", tablicaB[1](3))
x = 0
print (x, " + 3 =", tablicaA[1](3), "    1 + 3 = ",  tablicaB[1](3))
